hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ToOneAttributeMapping.java

```diff
@@ -169,101 +176,145 @@ public Fetch resolveCircularFetch(
-		final ForeignKeyDescriptor associationParentForeignKeyDescriptor = associationParent.getForeignKeyDescriptor();
-		if ( referencedModePart instanceof ToOneAttributeMapping
-				&& ( (ToOneAttributeMapping) referencedModePart ).getDeclaringType() == getPartMappingType() ) {
-			if ( this.foreignKeyDescriptor.getReferringTableExpression()
-					.equals( associationParentForeignKeyDescriptor.getReferringTableExpression() ) ) {
-				final SingleTableEntityPersister entityPersister = (SingleTableEntityPersister) getDeclaringType();
-				if ( associationParentForeignKeyDescriptor.getTargetTableExpression()
-						.equals( entityPersister.getTableName() ) ) {
-					final String[] identifierColumnNames = entityPersister.getIdentifierColumnNames();
-					if ( associationParentForeignKeyDescriptor.areTargetColumnNamesEqualsTo( identifierColumnNames ) ) {
-						return createBiDirectionalFetch( fetchablePath, fetchParent );
-					}
-					return null;
+					 */
+
+				if ( fetchablePath.getFullPath().endsWith( otherSideMappedBy ) ) {
+					return createCircularBiDirectionalFetch(
+							fetchablePath,
+							fetchParent,
+							parent.getParent(),
+							LockMode.READ
+					);
 				}
+			}
+			/*
+						class Child {
+							@OneToOne
+							private Mother mother;
+						}
 
+						class Mother {
+							@OneToOne
+							private Child stepMother;
+						}
+
+				We have a cirularity but it is not bidirectional
+			 */
+			if ( referringPrimaryKey ) {
+				final TableGroup parentTableGroup = creationState
+						.getSqlAstCreationState()
+						.getFromClauseAccess()
+						.getTableGroup( fetchParent.getNavigablePath() );
+				return new CircularFetchImpl(
+						getEntityMappingType(),
+						getTiming(),
+						fetchablePath,
+						fetchParent,
+						this,
+						fetchablePath,
+						foreignKeyDescriptor.createDomainResult( fetchablePath, parentTableGroup, creationState )
+				);
 			}
 		}
 		return null;
 	}
 
-	private Fetch createBiDirectionalFetch(NavigablePath fetchablePath, FetchParent fetchParent) {
-		final EntityResultGraphNode referencedEntityReference = resolveEntityGraphNode( fetchParent );
+	private String getOthrerSideMappedBy(
+			ModelPart modelPart,
+			NavigablePath parentOfParent,
+			DomainResultCreationState creationState) {
+		if ( modelPart instanceof ToOneAttributeMapping ) {
+			return ( (ToOneAttributeMapping) modelPart ).getMappedBy();
+		}
 
-		if ( referencedEntityReference == null ) {
-			throw new HibernateException(
-					"Could not locate entity-valued reference for circular path `" + fetchablePath + "`"
-			);
+		if ( modelPart instanceof EntityCollectionPart ) {
+			if ( parentOfParent.getFullPath().endsWith( EntityIdentifierMapping.ROLE_LOCAL_NAME ) ) {
+				parentOfParent = parentOfParent.getParent();
+			}
+			return ( (PluralAttributeMapping) creationState.resolveModelPart( parentOfParent ) ).getMappedBy();
 		}
```

Comment:
seems combining the above two if together in one if leads to better readability, though not a big deal